### ISSUE TITLE
Eagerloading has_one through association with references causing N+1 queries for nil

### ISSUE THREAD

**Issue by salmanasiddiqui** [rails/rails#21497]
rails 4.2.0
ruby 2.1.5

````
class Equipment < ActiveRecord::Base
  belongs_to :project
  has_one :portfolio, through: :project
end
````
````
class Project < ActiveRecord::Base
  belongs_to :portfolio
  has_many :equipments
end
````
````
class Portfolio < ActiveRecord::Base
  has_many :projects
  has_many :equipments, through: :projects
end
````

````
Equipment.includes(:project, :portfolio).references(:portfolio).each do |e|
  e.portfolio.try(:name)    #accessing has_one through association
end
````
The above is making multiple queries for Portfolio where project id is null.

though the below works fine
````
Equipment.includes(project: :portfolio).references(:portfolio).each do |e|
  e.project.try(:portfolio).try(:name)    #accessing has_one and has_one association
end
````

**Comment by codebaker95** [comment]
I don't seem to be able to replicate your problem. Could you show what exactly is your rails console showing when you make these requests? For me, it's this single request:

```
SQL (0.2ms)  SELECT "equipment"."id" AS t0_r0, "equipment"."name" AS t0_r1, 
"equipment"."project_id" AS t0_r2, "equipment"."created_at" AS t0_r3, "equipment"."updated_at" AS 
t0_r4, "projects"."id" AS t1_r0, "projects"."name" AS t1_r1, "projects"."created_at" AS t1_r2, 
"projects"."updated_at" AS t1_r3, "projects"."portfolio_id" AS t1_r4, "portfolios"."id" AS t2_r0, 
"portfolios"."name" AS t2_r1, "portfolios"."created_at" AS t2_r2, "portfolios"."updated_at" AS t2_r3 
FROM "equipment" LEFT OUTER JOIN "projects" ON "projects"."id" = "equipment"."project_id" LEFT 
OUTER JOIN "projects" "projects_equipment_join" ON "projects_equipment_join"."id" = 
"equipment"."project_id" LEFT OUTER JOIN "portfolios" ON "portfolios"."id" = 
"projects_equipment_join"."portfolio_id"
```

**Comment by salmanasiddiqui** [comment]
@codebaker95 
have you executed this loop?
````
Equipment.includes(:project, :portfolio).references(:portfolio).each do |e|
  e.portfolio.try(:name)
end
````

Notice ````e.portfolio.try(:name)````. I am accessing the has_one through association, and everytime there is an ````Equipment```` object where ````project_id```` is ````null````, it makes an extra query
 ````
Portfolio Load (0.3ms)  SELECT  "portfolios".* FROM "portfolios" INNER JOIN "projects" ON "portfolios"."id" = "projects"."portfolio_id" WHERE "projects"."id" = ? LIMIT 1  [["id", nil]]
````

**Comment by sgrif** [comment]
I'm also unable to reproduce. I created a script to test this, which you can find [here](https://gist.github.com/sgrif/f737b054f11a932e0880). I have checked both 4.2 and master, and in both cases it executes a single query properly. You should update your app to 4.2.4, which does not have this bug.

For the record, this is a dup of https://github.com/rails/rails/issues/18312 and was fixed in https://github.com/rails/rails/pull/18316. (We do have a search feature. :smile:)

**Comment by salmanasiddiqui** [comment]
@sgrif 
[here](https://github.com/salmanasiddiqui/rails-has-one-through) is a sample project. Please follow the short readme to reproduce the bug.

**Comment by sgrif** [comment]
Did you read my previous comment? Your gemfile in that example is pointed at 4.2.0. This was fixed in 4.2.3. Update your version of Rails.

**Comment by salmanasiddiqui** [comment]
yes I read your comment and suggestion to update my actual app to 4.2.4, but you also said that this issue is not reproducible on 4.2. That's why I shared that sample project with 4.2.

Thanks for your time
